ompi/mca/coll/acoll/coll_acoll_utils.h

```diff
@@ -0,0 +1,769 @@
+    (subc->socket_comm)->c_coll->coll_allreduce = coll_allreduce_soc;
+    (subc->socket_comm)->c_coll->coll_allgather = coll_allgather_soc;
+    (subc->socket_comm)->c_coll->coll_bcast = coll_bcast_soc;
+
+    /* For collectives where order is important (like gather, allgather),
+     * split based on ranks. This is optimal for global communicators with
+     * equal split among nodes, but suboptimal for other cases.
+     */
+    if (!subc->initialized) {
+        if (subc->num_nodes > 1) {
+            int node_size = (size + subc->num_nodes - 1) / subc->num_nodes;
+            int color = rank / node_size;
+            err = ompi_comm_split(comm, color, rank, &subc->local_r_comm, false);
+            if (MPI_SUCCESS != err) {
+                return err;
+            }
+        }
+        subc->derived_node_size = (size + subc->num_nodes - 1) / subc->num_nodes;
+    }
+
+    /* Restore originals */
+    (comm)->c_coll->coll_allreduce = coll_allreduce_org;
+    (comm)->c_coll->coll_allgather = coll_allgather_org;
+    (comm)->c_coll->coll_bcast = coll_bcast_org;
+
+    /* Init done */
+    subc->initialized = 1;
+    if (root != subc->prev_init_root) {
+        subc->num_root_change++;
+    }
+    subc->prev_init_root = root;
+
+    return err;
+}
+
+#ifdef HAVE_XPMEM_H
+static inline int mca_coll_acoll_xpmem_register(void *xpmem_apid, void *base, size_t size,
+                                                mca_rcache_base_registration_t *reg)
+{
+    struct xpmem_addr xpmem_addr;
+    xpmem_addr.apid = *((xpmem_apid_t *) xpmem_apid);
+    xpmem_addr.offset = (uintptr_t) base;
+    struct acoll_xpmem_rcache_reg_t *xpmem_reg = (struct acoll_xpmem_rcache_reg_t *) reg;
+    xpmem_reg->xpmem_vaddr = xpmem_attach(xpmem_addr, size, NULL);
+
+    if ((void *) -1 == xpmem_reg->xpmem_vaddr) {
+        return -1;
+    }
+    return 0;
+}
+
+static inline int mca_coll_acoll_xpmem_deregister(void *xpmem_apid,
+                                                  mca_rcache_base_registration_t *reg)
+{
+    int status = xpmem_detach(((struct acoll_xpmem_rcache_reg_t *) reg)->xpmem_vaddr);
+    return status;
+}
+#endif
+
+static inline int coll_acoll_init(mca_coll_base_module_t *module, ompi_communicator_t *comm,
+                                  coll_acoll_data_t *data)
+{
+    int size, ret = 0, rank, line;
+
+    mca_coll_acoll_module_t *acoll_module = (mca_coll_acoll_module_t *) module;
+    coll_acoll_subcomms_t *subc;
+    int cid = ompi_comm_get_local_cid(comm);
+    subc = &acoll_module->subc[cid];
+    if (subc->initialized_data) {
+        return ret;
+    }
+    subc->cid = cid;
+    data = (coll_acoll_data_t *) malloc(sizeof(coll_acoll_data_t));
+    if (NULL == data) {
+        line = __LINE__;
+        ret = OMPI_ERR_OUT_OF_RESOURCE;
+        goto error_hndl;
+    }
+    size = ompi_comm_size(comm);
+    rank = ompi_comm_rank(comm);
+    data->comm_size = size;
+
+#ifdef HAVE_XPMEM_H
```

Comment:
I don't see an include for the xpmem header. Where does it come from? Maybe that should be explicit if `HAVE_XPMEM_H` is defined?